### PR TITLE
enhm/globals

### LINKED ISSUE #164: Trace global variables better
Create a separate category specifically for global variables, and make sure that the annotation generators account for this when reviewing targets

### PR DESCRIPTION
**Pull request by bengsparks**
Closes #164

### CODE REVIEW

**Comment by bengsparks** [comment]
This implements the described issue. A refactor of the Tracer is definitely necessary, and can be merged into this branch prior to merging into develop.

File: tests/resource/typegen/glbls.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+from tracing import decorators
```
Comment (RecurvedBow) [inline tests/resource/typegen/glbls.py:1]: Remove this

File: tests/resource/typegen/glbls.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    sneaky_inside_scope = "TypeChange"
+
+
+def h():
+    # Reference existing global
+    global exists_outside_of_all_scopes
+    exists_outside_of_all_scopes = 5
+
+
+@decorators.trace
```
Comment (RecurvedBow) [inline tests/resource/typegen/glbls.py:26]: And this too

File: tests/tracing/test_tracer.py [hunk 3]
```diff
@@ -580,31 +607,74 @@ def test_if_tracer_traces_sample_function_with_inner_function_it_collects_correc
...
+def test_if_tracer_stops_no_trace_is_set(tracers: list[Tracer]):
+    for tracer in tracers:
+        tracer.stop_trace()
         assert sys.gettrace() is None
+
+
+import logging
+
+
+def test_if_tracer_finds_global(tracers: list[Tracer]):
```
Comment (RecurvedBow) [inline tests/tracing/test_tracer.py:640]: Remove the if in the name.

File: tests/typegen/strats/test_inline.py [hunk 4]
```diff
@@ -913,3 +915,142 @@ def visit_ImportFrom(self, node: cst.ImportFrom) -> bool | None:
...
+    print(exists_outside_of_all_scopes, sneaky_inside_scope)
+
+    # 5 TypeChange
+    h()
+    print(exists_outside_of_all_scopes, sneaky_inside_scope)
+
+
+if __name__ == "__main__":
+    main()
+"""
```
Comment (RecurvedBow) [inline tests/typegen/strats/test_inline.py:1043]: Update this so it matches resource.

File: typegen/strats/inline.py [hunk 5]
```diff
@@ -90,7 +91,9 @@ def __init__(self, module: str, relevant: pd.DataFrame) -> None:
         self.df.loc[mask, Column.VARTYPE] = "None"
 
         self._module = module
-        self._scope_stack: list[cst.FunctionDef | cst.ClassDef] = []
+        self._scope_stack: list[cst.FunctionDef | cst.ClassDef | cst.Global] = []
+
```
Comment (RecurvedBow) [inline typegen/strats/inline.py:95]: Globals don't have scopes. Fix this.

**Review by RecurvedBow — CHANGES_REQUESTED** [review]

File: typegen/strats/inline.py (hunk 5 above)
Comment (bengsparks) [reply typegen/strats/inline.py:95]: Simply removed. Globals are tracked separately; this is a remnant of the first approach

File: tests/resource/typegen/glbls.py (hunk 1 above)
Comment (bengsparks) [reply tests/resource/typegen/glbls.py:1]: Done

File: tests/tracing/test_tracer.py (hunk 3 above)
Comment (bengsparks) [reply tests/tracing/test_tracer.py]: Done

File: tests/resource/typegen/glbls.py (hunk 2 above)
Comment (bengsparks) [reply tests/resource/typegen/glbls.py]: Done

**Review by RecurvedBow — APPROVED** [review]

**Pull request merged**